### DISCUSSION TITLE
Detecting Collision Between Two Characters

### DISCUSSION THREAD

**Discussion by matthewbean** [Annoraaq/grid-engine#181] [Q&A]
Hi there, I really appreciate you creating this and making it much easier to create tile base games. The one thing I am running into that I cannot find an answer to, is how to allow collision between two characters and how to detect it. For example, I would like to have multiple NPC  characters walking around randomly as enemies, and I would like to detect them  colliding with the hero for damage and such. As of right now, they seemed to be just walking around and refuse to walk in the same  square is my character. Is there a way to deactivate this collision detection behavior and detect if there is a collision that happens?

**Comment by Annoraaq** [comment]
Hi @matthewbean,
I am glad you like the plugin and I hope I can help you out.

The way collision detection works in the plugin is that a tile/square/field is considered to be blocking if the position contains either a 'colliding' tile on any layer of the tilemap **OR** a character.

Since lately there is a config flag that allows you to disable collisions for a character completely. However, this means that this character will not collide with anything on the map. So it could walk through walls and so on.

So am I right with the assumption that you would like to have your enemy-NPCs to still collide with the tilemap but NOT with the player? If so, should they collide with each other (should an enemy collide with another enemy)?

Based on your demands I could add a flag that would allow you to specify collision detection more precise.

**Reply by matthewbean** [reply]
Hi there, thank you for responding. For my immediate usage, I would be OK removing their collusion completely. Their movement would be scripted completely, so I would be able to prevent them from colliding with anything in the environment. That would be a cool feature though, giving more granular control over what they can collide with. If that were a feature, I think it would be most useful if they were only able to collide with the player, not with each other. But for now, how would I go about disabling their collision completely, and what would be the best way to detect a collision with the player?

**Reply by Annoraaq** [reply]
To disable the collision for a character you can add the `collides: false` property like so:
```javascript
 const gridEngineConfig = {
    characters: [
      {
        id: "enemy",
        sprite: enemySprite,
        walkingAnimationMapping: 6,
        collides: false
      },
    ],
  };
  ```
  
  I initially forgot to add that to the docs, but just updated them.
  
  To your second question: _what would be the best way to detect a collision with the player?_
  
  One way would be to subscribe to the enemies `positionChangeFinished()` observable (depending on when you want the collision to happen you could instead subscribe to `positionChangeStarted()`) like so (untested):
  
```javascript
gridEngine.positionChangeFinished().subscribe(({charId, enterTile}) => {
  const playerPos = gridEngine.getPosition('player');
  if (charId !== 'player' && playerPos.x === enterTile.x && playerPos.y === enterTile.y) {
    // collision with player
  }
});
```

**Reply by matthewbean** [reply]
Awesome! I figured there was some thing that needed to go into the character object. But as you pointed out I could not find it in the documentation. Thank you for getting back to me, that should be good enough for me for now.

**Comment by splashsky** [comment]
Almost seems like collision layers would be the logical next step 🤔

**Reply by Annoraaq** [reply]
Can you give some more details on how these collision layers should work? It seems to be an interesting concept.

**Reply by lendsp** [reply]
> Almost seems like collision layers would be the logical next step 🤔

I'm testing this, I intend to have a terrestrial layer and a flying layer...
for now despite implementing the gridengine I do the collision control on the server...

var map = require ('./public/assets/cloud-city_collides.json');
layer1 = Object.values(map.layers[0])
layer2 = Object.values(map.layers[1])
layer3 = Object.values(map.layers[2])
layer4 = Object.values(map.layers[3])
layer1t=Object.values(layer1[0]);
layer2t=Object.values(layer2[0]);
layer3t=Object.values(layer3[0]);
layer4t=Object.values(layer4[0]);
function getLayer(cid,cmap,layer){
  layer1 = Object.values(map.layers[0]);
  layer2 = Object.values(map.layers[1]);
  layer3 = Object.values(map.layers[2]);
  layer1t=Object.values(layer1[0]);
  layer2t=Object.values(layer2[0]);
  layer3t=Object.values(layer3[0]);
  if (layer === 1) return layer1t[cid];
  if (layer === 2) return layer2t[cid];
  if (layer === 3) return layer3t[cid];
}

it's still pretty primitive, but that way I can check one layer at a time before moving the character

**Reply by Annoraaq** [reply]
I posted a solution here: https://github.com/Annoraaq/grid-engine/discussions/184#discussioncomment-1222052
Would that be a suitable option?
